src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,702 @@
+
+      // The SCC is not necessarily topologically sorted to have the supertypes
+      // come first. Fix that.
+      std::vector<HeapType> sccTypes(scc.begin(), scc.end());
+      auto deps = createSubtypeGraph(sccTypes);
+      auto permutation = *TopologicalOrders(deps).begin();
+      groups.back().group.resize(sccTypes.size());
+      for (size_t i = 0; i < sccTypes.size(); ++i) {
+        groups.back().group[i] = sccTypes[permutation[i]];
+      }
+      updateShape(index);
+    }
+
+    rewriteTypes(*module);
+  }
+
+  void initBrandOptions() {
+    // Initialize the field options for brand types lazily here to avoid
+    // depending on global constructor ordering.
+    [[maybe_unused]] static bool fieldsInitialized = []() {
+      BrandTypeIterator::initFieldOptions();
+      return true;
+    }();
+  }
+
+  void updateShape(size_t group) {
+    auto [it, inserted] =
+      groupShapeIndices.insert({RecGroupShape(groups[group].group), group});
+    if (inserted) {
+      // This shape was unique. We're done.
+      return;
+    }
+
+    // We have a conflict. There are five possibilities:
+    //
+    //   1. We are trying to insert the next permutation of an existing
+    //      equivalence class and have found that...
+    //
+    //     A. The next permutation is an automorphism of some previous
+    //        permutation of the same equivalence class.
+    //
+    //     B. The next permutation is equivalent to some other existing group
+    //        shape already in a different nontrivial equivalent class.
+    //
+    //     C. The next permutation is equivalent to some other existing group
+    //        shape not yet included in a nontrivial equivalence class.
+    //
+    //   2. We are inserting a new group shape as of yet unaffiliated with a
+    //      nontrival equivalence class and have found that...
+    //
+    //     A. It is the same shape as some group in an existing nontrival
+    //        equivalence class.
+    //
+    //     B. It is the same shape as some other group also not yet affiliated
+    //        with a nontrival equivalence class, so we have a new nontrivial
+    //        equivalence class.
+    //
+    // These possibilities are handled in order below.
+
+    size_t other = it->second;
+
+    auto& groupInfo = groups[group];
+    auto& otherInfo = groups[other];
+
+    size_t groupRep = equivalenceClasses.getRoot(group);
+    size_t otherRep = equivalenceClasses.getRoot(other);
+
+    // Case 1A: There is an automorphism. Skip the rest of the permutations,
+    // which will also be automorphic to previous permutations.
+    if (groupRep == otherRep) {
+      assert(groups[groupRep].classInfo);
+      auto& classInfo = *groups[groupRep].classInfo;
+
+      // Move to the next permutation after advancing the type brand to skip
+      // further automorphisms.
+      classInfo.advanceBrand();
+      classInfo.permute(groupInfo);
+
+      updateShape(group);
+      return;
```

Comment:
Is there no risk of getting a very deep stack due to this recursion? It seems like in principle there could be a conflict after adding the brand/permuting, and so forth. I'm not sure if it's worth the complexity, but maybe a comment?

Reply:
Yes, you could set up a pathological input with an arbitrarily deep stack by visiting each permution of a group separately and only then seeing a repeat group that causes the whole thing to collapse into a single equivalence class. A more naive approach would also risk long stretches of useless work as it iterates through permutations that are automorphic to previous permutations, but we eliminate that source of repetition by construction.

I don't think it will be too bad to switch to a worklist here, so I'll do that.